### PR TITLE
Uses importing module memory for host function calls

### LINKED ISSUE #195: Host functions called from another host function cannot know the right MemoryInstance
In the current implementation, host functions try to get the `MemoryInstance` of `HostFunctionCallContext` from `<function>.ModuleInstance.Memory` of the parent stack's (in interpreter) or of the function itself (in jit).
However, a host function doesn't have associated `MemoryInstance` nor wasm modules in the first place. So, a host function only finds `nil` as `MemoryInstance` when it's called from another host function. 

FYI, it looks wasmtime also doesn't provide the access to memory instances when a host function is called from another host function because there's no caller context in that case, but let me check.

https://github.com/tetratelabs/wazero/blob/a6cd1549ba093e35d09fe07cd1a049b70e798185/wasm/interpreter/interpreter.go#L516
https://github.com/tetratelabs/wazero/blob/a6cd1549ba093e35d09fe07cd1a049b70e798185/wasm/jit/engine.go#L390

This also prevents the WASI implementation from testing APIs by "import-and-exported-back" style, where tests call each WASI API by having a wasm module import a WASI API and re-export it to the host.

### PR DESCRIPTION
**Pull request by codefromthecrypt**
Fixes #195

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
I raised for review while I backfill unit tests

**Comment by mathetake** [comment]
cool!

File: tests/engine/adhoc_test.go [hunk 1]
```diff
@@ -201,31 +202,40 @@ func recursiveEntry(t *testing.T, newEngine func() wasm.Engine) {
...
 
-	// We should be able to call the exported add_int and it should add two ints.
-	results, _, err := store.CallFunction(ctx, "test", "add_int", uint64(12), uint64(30))
+	// We should be able to call the exported store_int and it should store two ints.
+	results, _, err := store.CallFunction(ctx, "test", "store_int", 1, math.MaxUint64)
 	require.NoError(t, err)
-	require.Equal(t, uint64(42), results[0])
+	require.Equal(t, wasi.ErrnoSuccess, wasi.Errno(results[0]))
+
+	require.Equal(t, []byte{0x0, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x0}, store.Memories[0].Buffer[0:10])
```
Comment (mathetake) [inline tests/engine/adhoc_test.go:238]: maybe comment on from what we have 0xff, 0xff

**Review by mathetake — APPROVED** [review]
Thank you for fixing (my) bug!

File: tests/engine/adhoc_test.go (hunk 1 above)
Comment (codefromthecrypt) [reply tests/engine/adhoc_test.go:238]: sure

**Comment by codefromthecrypt** [comment]
Changed to make WASI tests test the constants mentioned in docs (notably the import is now a constant instead of embedded godoc)

File: tests/engine/adhoc_test.go [hunk 2]
```diff
@@ -201,31 +202,41 @@ func recursiveEntry(t *testing.T, newEngine func() wasm.Engine) {
...
+		;; store_int is imported from the environment, but it's also exported back to the environment
+		(export "store_int" (func $store_int))
 		)`))
 	require.NoError(t, err)
 
 	store := wasm.NewStore(newEngine())
 
-	addInt := func(ctx *wasm.HostFunctionCallContext, x int32, y int32) int32 {
-		return x + y
+	storeInt := func(ctx *wasm.HostFunctionCallContext, offset uint32, val uint64) wasi.Errno {
```
Comment (mathetake) [inline tests/engine/adhoc_test.go:220]: I don't see any reason to use wasi.Errono.. simply use uint32?

Comment (codefromthecrypt) [reply tests/engine/adhoc_test.go:220]: ok sure

**Review by mathetake — APPROVED** [review]

**Comment by codefromthecrypt** [comment]
FYI @nullpo-head @r8d8 this PR drifts your PRs in a good way as testing the wasi functions is easier (since you can use a constant for the imports instead of copy/pasta of a %.wat file)

**Comment by codefromthecrypt** [comment]
Thanks for the review @mathetake and the hints on the issue and offline @nullpo-head

**Pull request merged**